File: SamplePresence/UIKit-ObjC/AuthViewController.m
#import "AuthViewController.h"
#import "DemoFlowController.h"
#import "FormFieldCell.h"

typedef NS_ENUM(NSInteger, AuthRow) {
    AuthRowServerURL = 0,
    AuthRowUserEmail = 1,
    AuthRowCount = 2
};

@interface AuthViewController ()
@property (nonatomic, weak) DemoFlowController *flowController;
@end

@implementation AuthViewController

- (instancetype)initWithFlowController:(DemoFlowController *)flowController {
    self = [super initWithStyle:UITableViewStyleInsetGrouped];
    if (self) {
        _flowController = flowController;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = @"Authenticate";
    [self.tableView registerClass:[FormFieldCell class] forCellReuseIdentifier:FormFieldCell.reuseIdentifier];
    self.navigationItem.rightBarButtonItem =
        [[UIBarButtonItem alloc] initWithTitle:@"Authenticate"
                                         style:UIBarButtonItemStyleDone
                                        target:self
                                        action:@selector(authenticateTapped)];
    [self bindFlowController];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return self.flowController.errorMessage == nil ? 1 : 2;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return section == 0 ? AuthRowCount : 1;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.section == 1) {
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"ErrorCell"];
        if (!cell) {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"ErrorCell"];
            cell.selectionStyle = UITableViewCellSelectionStyleNone;
            cell.textLabel.numberOfLines = 0;
        }
        cell.textLabel.text = self.flowController.errorMessage;
        cell.textLabel.textColor = UIColor.systemRedColor;
        return cell;
    }

    FormFieldCell *cell = [tableView dequeueReusableCellWithIdentifier:FormFieldCell.reuseIdentifier forIndexPath:indexPath];
    DemoSession *session = self.flowController.session;
    __weak typeof(self) weakSelf = self;

    if (indexPath.row == AuthRowServerURL) {
        [cell configureWithTitle:@"Server URL:"
                            text:session.serverURL
                     placeholder:@"http://127.0.0.1:8777"
                    keyboardType:UIKeyboardTypeURL];
        [cell.valueField addTarget:self action:@selector(serverURLChanged:) forControlEvents:UIControlEventEditingChanged];
    } else {
        [cell configureWithTitle:@"User Email:"
                            text:session.userEmail
                     placeholder:@"user@example.com"
                    keyboardType:UIKeyboardTypeEmailAddress];
        [cell.valueField addTarget:self action:@selector(userEmailChanged:) forControlEvents:UIControlEventEditingChanged];
    }

    (void)weakSelf;
    return cell;
}

- (void)serverURLChanged:(UITextField *)field {
    self.flowController.session.serverURL = field.text ?: @"";
}

- (void)userEmailChanged:(UITextField *)field {
    self.flowController.session.userEmail = field.text ?: @"";
}

- (void)authenticateTapped {
    [self.view endEditing:YES];
    [self.flowController authenticate];
}

- (void)bindFlowController {
    __weak typeof(self) weakSelf = self;
    self.flowController.onBusyChanged = ^(BOOL busy) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (!strongSelf) { return; }
        strongSelf.navigationItem.rightBarButtonItem.enabled = !busy;
        strongSelf.navigationItem.rightBarButtonItem.title = busy ? @"Authenticating…" : @"Authenticate";
    };
    self.flowController.onErrorChanged = ^(NSString * _Nullable errorMessage) {
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (!strongSelf) { return; }
        (void)errorMessage;
        [strongSelf.tableView reloadData];
    };
}

@end

File: SamplePresence/UIKit/AuthViewController.swift
import UIKit

final class AuthViewController: UITableViewController {
    private let flowController: DemoFlowController

    private enum Row: Int, CaseIterable {
        case serverURL
        case userEmail
    }

    init(flowController: DemoFlowController) {
        self.flowController = flowController
        super.init(style: .insetGrouped)
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        title = "Authenticate"
        tableView.register(FormFieldCell.self, forCellReuseIdentifier: FormFieldCell.reuseIdentifier)
        navigationItem.rightBarButtonItem = UIBarButtonItem(
            title: "Authenticate",
            style: .done,
            target: self,
            action: #selector(authenticateTapped)
        )
        bindFlowController()
    }

    override func numberOfSections(in tableView: UITableView) -> Int {
        flowController.errorMessage == nil ? 1 : 2
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if section == 0 {
            return Row.allCases.count
        }
        return 1
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if indexPath.section == 1 {
            let cell = UITableViewCell(style: .default, reuseIdentifier: "ErrorCell")
            cell.selectionStyle = .none
            cell.textLabel?.text = flowController.errorMessage
            cell.textLabel?.textColor = .systemRed
            cell.textLabel?.numberOfLines = 0
            return cell
        }

        guard let cell = tableView.dequeueReusableCell(withIdentifier: FormFieldCell.reuseIdentifier, for: indexPath) as? FormFieldCell,
              let row = Row(rawValue: indexPath.row) else {
            return UITableViewCell()
        }

        switch row {
        case .serverURL:
            cell.configure(
                title: "Server URL:",
                text: flowController.session.serverURL,
                placeholder: "http://127.0.0.1:8777"
            )
            cell.valueField.addAction(UIAction { [weak self] _ in
                self?.flowController.session.serverURL = cell.valueField.text ?? ""
            }, for: .editingChanged)
        case .userEmail:
            cell.configure(
                title: "User Email:",
                text: flowController.session.userEmail,
                placeholder: "user@example.com",
                keyboardType: .emailAddress
            )
            cell.valueField.addAction(UIAction { [weak self] _ in
                self?.flowController.session.userEmail = cell.valueField.text ?? ""
            }, for: .editingChanged)
        }

        return cell
    }

    @objc
    private func authenticateTapped() {
        view.endEditing(true)
        Task { await flowController.authenticate() }
    }

    private func bindFlowController() {
        flowController.onBusyChanged = { [weak self] isBusy in
            self?.navigationItem.rightBarButtonItem?.isEnabled = !isBusy
            self?.navigationItem.rightBarButtonItem?.title = isBusy ? "Authenticating…" : "Authenticate"
        }
        flowController.onErrorChanged = { [weak self] _ in
            self?.tableView.reloadData()
        }
    }
}
